lib/db/LocationSearchDAO.js

```diff
@@ -0,0 +1,226 @@
+    const limitApprox = limit - intersectionsExact.length;
+    if (limitApprox <= 0) {
+      return intersectionsExact;
+    }
+
+    /*
+     * For approximate matches, we use `pg_trgm` trigram support together with these features:
+     *
+     * - `f_word_similarity`: computes trigram similarity between query terms and tokens in
+     *   intersection names;
+     * - `f_feature_code`: same as for exact matches.
+     *
+     * The lower `pg_trgm.word_similarity_threshold` is chosen to balance precision and recall;
+     * PostgreSQL defaults this parameter to 0.6.
+     *
+     * This expression normalizes the query, removing stopwords and punctuation:
+     *
+     *     replace(replace(plainto_tsquery($(query))::text, '''', ''), ' &', '')
+     */
+    const sqlApprox = `
+SET pg_trgm.word_similarity_threshold = 0.3;
+WITH candidates AS (
+  SELECT
+    int_id,
+    word_similarity(
+      replace(replace(plainto_tsquery($(query))::text, '''', ''), ' &', ''),
+      intersec5
+    ) AS f_word_similarity,
+    elevatio9 AS f_feature_code
+  FROM location_search.centreline_intersection
+  WHERE replace(replace(plainto_tsquery($(query))::text, '''', ''), ' &', '') <% intersec5
+)
+SELECT
+  ci.int_id,
+  ci.intersec5,
+  ci.elevatio9,
+  ST_AsGeoJSON(ci.geom)::json AS geom,
+  f_word_similarity * 2 - abs(501245 - f_feature_code) / 100.0 AS score
+FROM candidates JOIN gis.centreline_intersection ci USING (int_id)
+ORDER BY score DESC, intersec5 ASC
+LIMIT $(limitApprox);`;
+    const rowsApprox = await db.manyOrNone(sqlApprox, {
+      limitApprox,
+      query,
+    });
+    const intersectionsApprox = rowsApprox.map(({ score, ...intersection }) => ({
+      score,
+      ...intersectionToFeature(intersection),
+    }));
+    return intersectionsExact.concat(intersectionsApprox);
+  }
+
+  /**
+   * Matches traffic signals using their PX / TSC number, as provided via special search
+   * terms like `px:1234`.
+   *
+   * Unlike {@link LocationSearchDAO.intersectionSuggestions}, this method doesn't really
+   * "suggest" results - it does an exact match on the PX / TSC number.
+   *
+   * @param {number} px - PX / TSC number
+   * @returns {Array<Object>} - array containing the centreline feature corresponding
+   * to the given signal if found, empty array otherwise
+   */
+  static async trafficSignalSuggestions(px) {
+    const sql = `
+SELECT
+  ci.int_id,
+  ci.intersec5,
+  ci.elevatio9,
+  ST_AsGeoJSON(ci.geom)::json AS geom
+FROM location_search.traffic_signal ts
+JOIN gis.centreline_intersection ci ON ts.centreline_id = ci.int_id
+WHERE ts.px = $(px)`;
+    const rows = await db.manyOrNone(sql, { px });
+    return rows.map(intersectionToFeature);
+  }
+
+  static async getSuggestions(query, limit) {
+    const queryNormalized = query.trim();
+
+    const match = REGEX_SPECIAL_TERM.exec(queryNormalized);
+    if (match !== null) {
+      const value = parseInt(match[2], 10);
+      if (Number.isNaN(value)) {
+        return [];
```

Comment:
eventually it would be nice to have error states on this, but we're not quite there yet